src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReferenceDataflowSource.java

```diff
@@ -0,0 +1,69 @@
+package org.broadinstitute.hellbender.engine.dataflow.datasources;
+
+import com.google.cloud.dataflow.sdk.options.PipelineOptions;
+import com.google.cloud.dataflow.sdk.transforms.SerializableFunction;
+import htsjdk.samtools.reference.ReferenceSequenceFileFactory;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.dataflow.BucketUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.reference.ReferenceBases;
+
+import java.io.IOException;
+import java.io.Serializable;
+
+/**
+ * Class to load a reference sequence from the Google Genomics API, or a file stored on HDFS or locally.
+ *
+ * This class needs to be mocked, so it cannot be declared final.
+ */
+public class ReferenceDataflowSource implements ReferenceSource, Serializable {
+    private static final long serialVersionUID = 1L;
+
+    private ReferenceSource referenceSource;
+    private SerializableFunction<GATKRead, SimpleInterval> referenceWindowFunction;
+
+    /**
+     * @param pipelineOptions the pipeline options; must be GCSOptions if using the Google Genomics API
+     * @param reference the name of the reference (if using the Google Genomics API), or a path to the reference file
+     * @param referenceWindowFunction the custom reference window function used to map reads to desired reference bases
+     */
+    public ReferenceDataflowSource(final PipelineOptions pipelineOptions, final String reference,
+                                   final SerializableFunction<GATKRead, SimpleInterval> referenceWindowFunction) {
+        if (isFasta(reference)) {
+            if (BucketUtils.isHadoopUrl(reference)) {
+                referenceSource = new ReferenceHadoopSource(reference);
+            } else {
+                referenceSource = new ReferenceFileSource(reference);
+            }
+        } else { // use the Google Genomics API
```

Comment:
Thanks for taking a look @davidadamsphd. What do you think the condition for this should be? The `reference` is an arbitrary ID (e.g. EOSt9JOVhp3jkwE), so there's no good way of testing for that case. We might check if the `pipelineOptions` can be converted to `GCSOptions` and that the necessary credentials are set, but that is done by `RefAPISource` anyway, which is the natural place to check it.